omod/src/main/java/org/openmrs/module/referenceapplication/WebComponentRegistrar.java

```diff
@@ -0,0 +1,42 @@
+/**
+ * The contents of this file are subject to the OpenMRS Public License
+ * Version 1.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
+ * http://license.openmrs.org
+ *
+ * Software distributed under the License is distributed on an "AS IS"
+ * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
+ * License for the specific language governing rights and limitations
+ * under the License.
+ *
+ * Copyright (C) OpenMRS, LLC.  All Rights Reserved.
+ */
+package org.openmrs.module.referenceapplication;
+
+import org.openmrs.module.referenceapplication.filter.RequireLoginLocationFilter;
+import org.springframework.stereotype.Component;
+import org.springframework.web.context.ServletContextAware;
+
+import javax.servlet.FilterRegistration;
+import javax.servlet.ServletContext;
+
+/**
+ * Dynamically registers web components like {@link javax.servlet.Filter},
+ * {@link javax.servlet.Servlet} etc.
+ */
+@Component(ReferenceApplicationConstants.MODULE_ID + "WebComponentRegistrar")
+public class WebComponentRegistrar implements ServletContextAware {
+	
+	private static final String FILTER_NAME = "requireLoginLocationFilter";
+	
+	@Override
+	public void setServletContext(ServletContext servletContext) {
+		//Don't re-register the filter on application context refresh because we already did.
```

Comment:
When the application context is refreshed, the openmrs class loader is destroyed. https://github.com/openmrs/openmrs-core/blob/2.2.0/api/src/main/java/org/openmrs/module/ModuleUtil.java#L882
What happens to the filter which was loaded by this class loader?

Reply:
@dkayiwa what about this then, this is wrong too in your opinion? https://github.com/openmrs/openmrs-module-legacyui/blob/521d14ff2edc106b997b50964431aab32b9c67c3/omod/src/main/java/org/openmrs/web/WebComponentRegistrar.java#L29-L65

Reply:
Yes. It does not work after the spring application context is refreshed.

Reply:
Hmmm, I tested and they do work actually, legacy UI module has been in use with this for years, if it was the case it should have been reported, Dimitri can you confirm if they don't work on your side too?

Reply:
@dkayiwa here is what I did in order to test @wluyima's new servlet filter (so using his fork of the Ref App module), as a super user:

1. I stopped a module that would trigger a context refresh, I used Attachments since it doesn't require anything.
1. I logged out. The two-screen login was still working.
1. I logged in again.
1. After logging in again I restarted Attachments.
1. I logged out. The two-screen login was still working.
1. I logged in again.

@dkayiwa clearly here the servlet filter survived multiple refreshes of the context.
Have you tried this on your end using the fork of @wluyima? You need to set any value to this GP for the two-screen login to be turned on: `"referenceapplication.locationUserPropertyName"`, as long as it's not blank/empty.

Reply:
That confirms a memory leak. Each time we have an application context refresh, all module class loaders are destroyed and created afresh. So the class loader for the reference application module sticks around because the servlet registration framework keeps holding onto a class that it loaded. As a result, we end up with multiple copies instead of one class loader for the same module. This memory leak does not happen if we use our openmrs custom filter registration in the config.xml file.

The above being said, am fine with merging this pull request, since in production, we recommend a servlet container restart for actions (like module start, stop, install, etc) that would result into the application context refresh.

Reply:
Yes when you register the filter this way, the filter actually sticks around because that's just the way true filters and servlets are handled by a servlet container which what our 'fake' filters are not  subject to and I'm aware of that, and just as you just confirmed it's not actually practical to restart the reference application module and in fact pretty much over 90% of modules in the RA, so you are always going to instead restart the OpenMRS instance which is why I use this approach.

But for the sake of moving on with this, I'm fine with to switch to the old way since anyways I'm not using the features that are limited by our synthetic module filters.